add new spec to references pages (#6190)
fishtown-analytics/docs.getdbt.com: website/docs/reference/snapshot-configs.md

-->

## Available configurations

> ### Snapshot-specific configurations

Reviewer: We have a section header like this for [models](https://docs.getdbt.com/reference/model-configs), [seeds](https://docs.getdbt.com/reference/seed-configs), etc. so for consistency, I think we should leave this as-is for now and not remove it.

And then if we want to update it, do it for all resource types at the same time.

<ConfigResource meta={frontMatter.meta} />

<VersionBlock lastVersion="1.8">

:::info Use the latest snapshot syntax

In Versionless and dbt v1.9 and later, snapshots are defined in an updated syntax using a YAML file within your `snapshots/` directory (as defined by the [`snapshot-paths` config](/reference/project-configs/snapshot-paths)). For faster and more efficient management, consider the updated snapshot YAML syntax, [available in Versionless](/docs/dbt-versions/versionless-cloud) or [dbt Core v1.9 and later](/docs/dbt-versions/core).

> :::

Reviewer: We've got a couple of these -- would we ever want a snippet for this info call out?

Author: good callout!

</VersionBlock>

<Tabs
  groupId="config-languages"
  defaultValue="project-yaml"
  values={[
    { label: 'Project file', value: 'project-yaml', },

>     { label: 'YAML file', value: 'property-yaml', },

Reviewer: Note for something for us to consider later:
- We refer to this a _lot_ of different ways across our docs: schema.yml, properties.yml, <file_name.yml>, YAML, Property file, Properties files, etc. Would be awesome for us to standardize to the extent possible!

</TabItem>

<TabItem value="config-resource">

> <VersionBlock firsttVersion="1.9">

Reviewer: ```suggestion
<VersionBlock firstVersion="1.9">
```

> Configurations can be applied to snapshots using the more performant [YAML syntax](/docs/build/snapshots), aavilable in Versionless and dbt v1.9 and higher, in the the `snapshot` directory file.

Reviewer: ```suggestion
Configurations can be applied to snapshots using [YAML syntax](/docs/build/snapshots), available in Versionless and dbt v1.9 and higher, in the the `snapshot` directory file.
```

```

</File>
</VersionBlock>

</TabItem>

<TabItem value="config">

> <VersionBlock firsttVersion="1.9">

Reviewer: ```suggestion
<VersionBlock firstVersion="1.9">
```

Author: great catch!

> Configurations can be applied to snapshots using the more performant [YAML syntax](/docs/build/snapshots), aavilable in Versionless and dbt v1.9 and higher, in the the `snapshot` directory file.

Reviewer: ```suggestion
Configurations can be applied to snapshots using [YAML syntax](/docs/build/snapshots), available in Versionless and dbt v1.9 and higher, in the the `snapshot` directory file.
```

Author: omg great catch!

```

</VersionBlock>

</TabItem>

</Tabs>

## Configuring snapshots
Snapshots can be configured in multiple ways:

<VersionBlock firstVersion="1.9">

1. Defined in YAML files using a `config` [resource property](/reference/model-properties), typically in your [snapshots directory](/reference/project-configs/snapshot-paths) (available in [Versionless](/docs/dbt-versions/versionless-cloud) or and dbt Core v1.9 and higher).
2. From the `dbt_project.yml` file, under the `snapshots:` key. To apply a configuration to a snapshot, or directory of snapshots, define the resource path as nested dictionary keys.
</VersionBlock>

<VersionBlock lastVersion="1.8">

1. Defined in YAML files using a `config` [resource property](/reference/model-properties), typically in your [snapshots directory](/reference/project-configs/snapshot-paths) (available in [Versionless](/docs/dbt-versions/versionless-cloud) or and dbt Core v1.9 and higher).

> 2. Using a `config` block within a snapshot

Reviewer: ```suggestion
2. Using a `config` block within a snapshot defined in Jinja SQL
```

3. From the `dbt_project.yml` file, under the `snapshots:` key. To apply a configuration to a snapshot, or directory of snapshots, define the resource path as nested dictionary keys.

Note that in Versionless and dbt v1.9 and later, snapshots are defined in an updated syntax using a YAML file within your `snapshots/` directory (as defined by the [`snapshot-paths` config](/reference/project-configs/snapshot-paths)). For faster and more efficient management, consider the updated snapshot YAML syntax, [available in Versionless](/docs/dbt-versions/versionless-cloud) or [dbt Core v1.9 and later](/docs/dbt-versions/core).

</VersionBlock>

> Snapshot configurations are applied hierarchically in the order above.

Reviewer: We [don't yet](https://github.com/dbt-labs/docs.getdbt.com/issues/4371) have standardized language and style that communicates the precedence order. In the meantime, maybe a slight modifier here to clarify:
```suggestion
Snapshot configurations are applied hierarchically in the order above with higher taking precedence.
```

I just now added snapshot configs to this list: https://github.com/dbt-labs/docs.getdbt.com/issues/4371

       config:
         unique_key: id
         strategy: timestamp
         updated_at: updated_at
         persist_docs:
           relation: true
           columns: true
    ```
    </File>
   Pro-tip: Use sources in snapshots: `select * from {{ source('jaffle_shop', 'orders') }}`
    </VersionBlock>

  You can also use the full resource path (including the project name, and subdirectories) to configure an individual snapshot from your `dbt_project.yml` file.

  For a project named `jaffle_shop`, with a snapshot file within the `snapshots/postgres_app/` directory, where the snapshot is named `orders_snapshot` (as above), this would look like:

    <File name='dbt_project.yml'>

    ```yml
    snapshots:
      jaffle_shop:
        postgres_app:
          orders_snapshot:
            +unique_key: id
            +strategy: timestamp
            +updated_at: updated_at
    ```

    </File>

  You can also define some common configs in a snapshot's `config` block. We don't recommend this for a snapshot's required configuration, however.

    <File name='dbt_project.yml'>

    ```yml
    version: 2

    snapshots:
      - name: orders_snapshot

>         config:
>           persist_docs:
>             relation: true
>             columns: true

Reviewer: I know this was just rearranged from pre-existing documentation, but I'm not sure if it actually works!

Per docs [here](https://docs.getdbt.com/reference/resource-configs/persist_docs), I think it should be the following (but I didn't try it to confirm or deny!):
```suggestion
        +persist_docs:
          relation: true
          columns: true
```

## Configuring snapshots
Snapshots can be configured in multiple ways:

> 1. Defined in YAML files, typically in your snapshots directory (available in [Versionless](/docs/dbt-versions/versionless-cloud) or and dbt Core v1.9 and higher).

Author: ```suggestion
1. Defined in YAML files, typically in your [snapshots directory](/reference/project-configs/snapshot-paths) (available in [Versionless](/docs/dbt-versions/versionless-cloud) or and dbt Core v1.9 and higher).
```

## Configuring snapshots

> Snapshots can be configured in one of three ways:

Reviewer: It says "three ways", so maybe we should generalize it like this?

```suggestion
Snapshots can be configured in multiple ways:
```

1. Defined in YAML files, typically in your snapshots directory (available in [Versionless](/docs/dbt-versions/versionless-cloud) or and dbt Core v1.9 and higher).
2. Using a `config` block within a snapshot

> 3. Using a `config` [resource property](/reference/model-properties) in a `.yml` file

Author: @dbeatty10 is the hierarchy correct here?

Reviewer: The hierarchy looks correct to me! 👍

For the 1st item, you could link to the [snapshots directory](https://docs.getdbt.com/reference/project-configs/snapshot-paths) or [`snapshot-paths`](https://docs.getdbt.com/reference/project-configs/snapshot-paths).
